logic/LogicManager.java

```diff
@@ -130,34 +130,45 @@ private Rule fromStructure(RuleStructure ruleStructure) {
     }
 
     private void validateCyclesThroughNegations(ConceptManager conceptMgr, LogicManager logicMgr) {
+        // Problem: When a derivation of an inferred concept contains its own negation.
+        // Tracking this at runtime is impractical - statically detect & disallow.
+        // Detect: A cycle which passes through atleast one negation.
+        //      Equivalently - a cycle starting from a negated concept in a rule.
         Set<Rule> negationRulesTriggeringRules = logicMgr.rulesWithNegations()
                 .filter(rule -> !rule.condition().negatedConcludablesTriggeringRules(conceptMgr, logicMgr).isEmpty())
                 .toSet();
 
         for (Rule negationRule : negationRulesTriggeringRules) {
             Map<Rule, RuleDependency> visitedDependentRules = new HashMap<>();
-            visitedDependentRules.put(negationRule, RuleDependency.of(negationRule, null));
-            List<RuleDependency> frontier = new LinkedList<>(ruleDependencies(negationRule, conceptMgr, logicMgr));
-            RuleDependency dependency;
+            LinkedList<RuleDependency> frontier = new LinkedList<>(negatedRuleDependencies(negationRule, conceptMgr, logicMgr));
             while (!frontier.isEmpty()) {
-                dependency = frontier.remove(0);
+                RuleDependency dependency = frontier.removeFirst();
+                visitedDependentRules.put(dependency.recursiveRule, dependency);
                 if (negationRule.equals(dependency.recursiveRule)) {
                     List<Rule> cycle = findCycle(dependency, visitedDependentRules);
                     String readableCycle = cycle.stream().map(Rule::getLabel).collect(Collectors.joining(" -> \n", "\n", "\n"));
                     throw TypeDBException.of(CONTRADICTORY_RULE_CYCLE, readableCycle);
                 } else {
-                    visitedDependentRules.put(dependency.recursiveRule, dependency);
                     Set<RuleDependency> recursive = ruleDependencies(dependency.recursiveRule, conceptMgr, logicMgr);
-                    recursive.removeAll(visitedDependentRules.values());
-                    frontier.addAll(recursive);
+                    iterate(recursive)
+                            .filter(rule -> !visitedDependentRules.containsKey(rule.recursiveRule))
+                            .forEachRemaining(ruleDependency -> {
+                                frontier.add(ruleDependency);
+                            });
```

Comment:
you can use `.forEachremaining(frontier::add)` here - if you remove the `{}`, intellij will suggest it for you!